KAFKA-17606: Include Rat errors in GitHub workflow summary

LoganZhuZzz opened the pull request:
In the current GitHub workflow, it has become difficult to locate Rat errors when running `check -x test`. To improve readability and make it easier to identify Rat errors, we need to include them in the workflow summary.

## Key Modifications
- Added a call to `rat.py` in the `build.yml` file to annotate Rat errors after the `Compile and validate` step.
- Ensured that the annotation for Rat reports will execute **only when the `Compile and validate` step fails.** 

## Purpose
- Enhance the usability of the GitHub workflow, allowing developers to more easily identify and fix Rat errors.
- Improve workflow efficiency by reducing the time spent manually searching for errors.

## Notes
- **Please ensure that all relevant workflow tests are run before merging this PR to confirm that the new feature operates correctly.**

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

LoganZhuZzz committed:
KAFKA-17606: Include Rat errors in GitHub workflow summary

LoganZhuZzz commented:
@chia7712 please take a look,thanks!

mumrah commented on .github/workflows/build.yml:
```diff
@@ -68,6 +68,11 @@ jobs:
           ./gradlew --build-cache --info  \
           ${{ inputs.is-public-fork && '--no-scan' || '--scan' }} \
           check -x test
+      - name: Annotate Rat errors
```
Can you move this down with the checkstyle step?

mumrah commented on .github/workflows/build.yml:
```diff
@@ -68,6 +68,11 @@ jobs:
           ${{ inputs.is-public-fork && '--no-scan' || '--scan' }} \
           check -x test
+      - name: Annotate Rat errors
+        if: ${{ failure() }}
```
We should only run this on java 21 to avoid duplicate reports from the parallel builds

mumrah reviewed (commented):
Thanks for the patch @LoganZhuZzz! 

Left a few minor comments inline. Can you temporarily add a couple of files without a license so we can see the script working in this PR?

LoganZhuZzz committed:
Refactor: Move 'Annotate Rat errors' step below 'Annotate checkstyle errors' and restrict to Java 21 builds

- Moved the 'Annotate Rat errors' step to follow the 'Annotate checkstyle errors' step, ensuring both error annotation steps are grouped for better workflow clarity.
- Added a conditional check to only run the 'Annotate Rat errors' step on Java 21 to prevent duplicate reports across parallel builds, in line with the checkstyle step behavior.
- Temporarily added test files without licenses to validate that unapproved licenses are correctly summarized in the GitHub workflow, per feedback.

LoganZhuZzz commented:
> Thanks for the patch @LoganZhuZzz!
> 
> Left a few minor comments inline. Can you temporarily add a couple of files without a license so we can see the script working in this PR?

Thank you for the feedback on the patch @mumrah 
I have made the following changes:

1. Temporarily added test files without licenses to validate the unapproved licenses summary in the GitHub workflow.
2. Moved the "Annotate Rat errors" step below the "Annotate checkstyle errors" step for better clarity.
3. Added a version check to ensure annotation steps run only on Java 21 to prevent duplicate reports.

Please let me know if there’s anything else you would like me to address!

LoganZhuZzz commented:
![1727412847728](https://github.com/user-attachments/assets/c4ee10cf-bb9d-431a-b91d-50cbaf15670c)

This pull request successfully integrates the annotation of Rat errors into the GitHub workflow summary.

LoganZhuZzz committed:
feat: Enhance Rat integration and error handling

- Added `ignoreRatFailure` parameter to `build.gradle` to control the `failOnError` behavior of the Rat task.
- Updated `build.yml` to pass the `ignoreRatFailure` parameter when invoking `check -x test`.
- Modified `rat.py` to exit with code 1 if unapproved licenses are found, ensuring that the workflow failure links directly to the Rat output for improved traceability.

This change facilitates better handling of Rat reports within the CI/CD pipeline.

LoganZhuZzz commented:
> Thanks @LoganZhuZzz, the script output looks good. Can you try the following for me:
> 
> 1. Add a `ignoreRatFailure` parameter to build.gradle that controls `rat { failOnError }`
> 2. Pass in this param in build.yml when we invoke `check -x test`
> 3. Modify rat.py to exit 1 if it finds any problems
> 
> This way, the link to the failed workflow will take us to the Rat output rather than the first failing step.
> 
> Also, could you include "::notice" output in the script similar to this https://github.com/apache/kafka/blob/trunk/.github/scripts/checkstyle.py#L64? This is how we actually create the annotations. Note that the line number starts with 1. https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/workflow-commands-for-github-actions#setting-a-notice-message

Thanks for the guidance. @mumrah 
I have fixed the above issues, If you notice any areas for improvement or any bad practices, please let me know.

mumrah commented on build.gradle:
```diff
@@ -210,6 +210,10 @@ def determineCommitId() {
 
 if (repo != null) {
+  ext {
+    ignoreRatFailure = project.hasProperty('ignoreRatFailure') ? project.ignoreRatFailure.toBoolean() : false
```
Can you relocate this with other property handling? (Along with userMaxForks, etc)

mumrah commented on .github/workflows/build.yml:
```diff
@@ -67,7 +67,7 @@ jobs:
           ./gradlew --build-cache --info  \
           ${{ inputs.is-public-fork && '--no-scan' || '--scan' }} \
-          check -x test
+          check -x test  -PignoreRatFailure=true
```
Can you move the property before the task on a separate line?

mumrah commented on .github/scripts/rat.py:
```diff
@@ -0,0 +1,91 @@
+            rel_path = os.path.relpath(file, workspace_path)
+            print(f"::notice file={rel_path},title=Unapproved License::File with unapproved license")
+
+    exit(1 if total_unapproved_licenses > 0 else 0)
```
Can you make this a regular if/else rather than inlined?

mumrah reviewed (commented):
Thanks for the updates @LoganZhuZzz! Just a few small comments. 

Can you try this out on your fork and include a link to the results?

LoganZhuZzz committed:
fix: Resolve issues with Gradle configuration

- Moved the `ignoreRatFailure` property to a separate line before `check -x test` for clarity.
- Added a comment for the `-PignoreRatFailure=true` flag to improve documentation.
- Relocated `ext` block configuration for better management alongside
  other properties

LoganZhuZzz committed:
Fix: Correct YAML syntax issue caused by indentation error

LoganZhuZzz commented:
Hi David @mumrah ! Here are the results on my fork:

### Compile and Validate Results
<img width="1080" alt="b8d38b77eb24bd5773c4949083d5e2c" src="https://github.com/user-attachments/assets/412a29f7-2406-4aea-b3ff-bb6a54d7fd61">

### Annotate Rat Errors Results
<img width="1080" alt="09c6fe160eb0abde9296ce9920b0fb3" src="https://github.com/user-attachments/assets/70a6538e-5996-4939-bb6a-662795eefacc">

[Link to Results](https://github.com/LoganZhuZzz/kafka/actions/runs/11091913918/job/30816286950)

---

### Questions

1. After modifying the `Annotate Rat errors` step from `if: ${{ failure() && matrix.java == '21' }}` to `if: ${{ matrix.java == '21' }}`, the `Annotate Rat errors` step was able to execute. Is this behavior in line with our expectations?
2. Do the results from `Annotate Rat errors` meet our expectations?

mumrah commented:
> When the ignoreRatFailure parameter is set to true, the failOnError of the Gradle RAT task is set to false.This causes the failure() condition in the Annotate Rat errors step of the workflow (build.yml) to evaluate as false, preventing the Annotate Rat errors step from being executed. Does this align with what we expect?

I see what you mean. In this case, let's do the same as we do for checkstyle and let RAT failures fail the build. Then, we will hit the `failure()` condition as expected. To do this, I think we just need to remove the `ignoreRatFailure` property.

LoganZhuZzz committed:
fix: Remove `ignoreRatFailure` property to allow RAT failures to fail the build

This change aligns the behavior of the RAT task with the checkstyle task, ensuring that any RAT failures will trigger the `failure()` condition in the workflow. This should enable the `Annotate Rat errors step` to execute as expected.

LoganZhuZzz commented:
Now, it meets our expectations. Any suggestions? @mumrah 

## Output Results
<img width="1080" alt="eb639b877fc14ba94f172149b6e13ce" src="https://github.com/user-attachments/assets/89751a5c-0b6a-4ebe-9e2d-335d5b274ab6">
<img width="1080" alt="edd5a162cc733152aad8930c8a35890" src="https://github.com/user-attachments/assets/001320be-3e53-4115-abe7-817ddb5da39d">

[Link to Results](https://github.com/apache/kafka/actions/runs/11098437481/job/30831116931?pr=17280)

mumrah commented:
@LoganZhuZzz, looks great! Please revert the temporary files and I'll merge this.

LoganZhuZzz committed:
revert: Revert temporary files

mumrah reviewed (approved):
LGTM, thanks @LoganZhuZzz!

edit: Also, thanks for the well written and well formatted PR :)
